docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+- No events are emitted on the non-main thread.
+- Less lazy when cancelling is possible (i.e. we do more work on the main thread before submitting the async task).
+- Splits up slicing logic into pre/post data reading, making program flow harder to follow.
+- Does not address goal 2.
+    
+### Use `QThread` and similar utilities instead of `concurrent.futures`
+
+- Standard way for plugins to support long running operations.
+- Can track progress and allow more opportunity for cancellation with `yielded` signal.
+- Can easily process done callback (which might update Qt widgets) on main thread.
+- Need to define our own task queue to achieve lazy slicing.
+- Need to connect a `QObject`, which ties our core to Qt, unless the code that controls threads does not live in core.
+    
+### Use `asyncio` package instead of `concurrent.futures`
+
+- May improve general readability of async code for some.
+- Mostly syntactic sugar on top of `concurrent.futures`.
+- Likely need an `asyncio` event loop distinct from Qt's main event loop, which could be confusing and cause issues.
+
+
+## Discussion
+
+- [Initial announcement on Zulip](https://napari.zulipchat.com/#narrow/stream/296574-working-group-architecture/topic/Async.20slicing.20project).
+    - Consider (re)sampling instead of slicing as the name for the operation discussed here.  
+- [Problems with `NAPARI_ASYNC=1`](https://forum.image.sc/t/even-with-napari-async-1-data-loading-is-blocking-the-ui-thread/68097/4)
+    - The existing experimental async code doesn't handle some seemingly simple usage.
+- [Remove slice state from layer](https://github.com/napari/napari/issues/4682)
+    - Feature request to remove slice/render state from the layer types.
+    - Motivated by creating multiple slices of the same layers for multiple canvases.
+- [Draft PR on andy-sweet's fork](https://github.com/andy-sweet/napari/pull/16)
+    - Initial feedback before sharing more widely.
+    
+### Open questions
+
+- Should `Dims.current_step` represent the last slice position request or the last slice response?
+    - With sync slicing, there is no distinction.
+    - If it represents the last slice response, then what should we connect the sliders to?
+    - Similar question for `corner_pixels` and others.
+
+- Which new classes, methods, and attributes should be private?
+    - Maybe all of them in the short term, so that we can wait for things to mature before committing to a public API.
+
+- Does this design work well with slicing operations other than slider movements?
+    - We mostly use moving the slider as a driving example here, but there are other things that cause slicing.
+    - E.g. changing the display dimensionality, panning, zooming, explicit call to `refresh`.
+    - If not, could we easily change it to do so?
+
+- Should we maintain the synchronous behavior of `refresh` and `set_view_slice`?
+
+- Should we invert the current design and submit async tasks within each vispy layer?
```

Comment:
This would prevent us from easily re-using this logic with another backend in the future.

Reply:
+1 vispy independent

Reply:
> This would prevent us from easily re-using this logic with another backend in the future.

Not necessarily. Most of the slicing logic proposed here is defined in the layers (in particular `Layer._get_slice`) and in `LayerSlicer` and this idea wouldn't change that. I haven't done a great job of explaining what the alternative is, so I'll probably update this text with some pseudocode and list it as an alternative considered.